### PR TITLE
[compiler][eslint] remove compilationMode override; report bailouts on first line

### PR DESCRIPTION
**Pull request by mofeiZ**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #30423
* #30394

---

Before:
- observe that both functions are compiled, even though `useFoo` is not defined using [component syntax](https://flow.org/en/docs/react/component-syntax/)
- eslint diagnostic highlights entire function body

https://github.com/user-attachments/assets/e4a26fa9-cebb-4a44-8789-4810a015a67b

After:
- `compilationMode` config is respected
- eslint diagnostic highlights only first line

https://github.com/user-attachments/assets/eb3f9b1a-b204-4f10-a920-d6fd6b821adc

### CODE REVIEW

File: compiler/packages/eslint-plugin-react-compiler/src/rules/ReactCompilerRule.ts [hunk 1]
```diff
@@ -100,7 +100,6 @@ function makeSuggestions(
 
 const COMPILER_OPTIONS: Partial<PluginOptions> = {
   noEmit: true,
-  compilationMode: "infer",
```
Comment (mofeiZ) [inline compiler/packages/eslint-plugin-react-compiler/src/rules/ReactCompilerRule.ts:103]: Note that `infer` is already the `compilationMode` default, so this override only affects user-specified `compilationMode`s (by overriding all other options to `infer`).

File: compiler/packages/eslint-plugin-react-compiler/src/rules/ReactCompilerRule.ts [hunk 2]
```diff
@@ -161,9 +160,16 @@ const rule: Rule.RuleModule = {
               detail.loc != null && typeof detail.loc !== "symbol"
                 ? ` (@:${detail.loc.start.line}:${detail.loc.start.column})`
                 : "";
+            const firstLineLoc = {
+              start: event.fnLoc.start,
+              end: {
+                line: event.fnLoc.start.line,
+                column: 10e3,
+              },
+            };
```
Comment (mofeiZ) [inline compiler/packages/eslint-plugin-react-compiler/src/rules/ReactCompilerRule.ts:169]: Slightly hacky, but the alternative is to attach another loc object to events (e.g. function identifier, parameters, etc). This felt cleaner, as not all functions we compile have id or param nodes.

Comment (mvitousek) [reply compiler/packages/eslint-plugin-react-compiler/src/rules/ReactCompilerRule.ts:169]: Worth noting that in Flow we have a `sig_loc` on functions, which is the location of, effectively, everything in a function other than its body (`function` keyword, name, type parameters, parameters including parentheses, => for arrow functions, return type annotation). This has been a useful thing for us to have in a lot of error messages and the compiler might benefit from something similar.

**Review by mvitousek — APPROVED** [review]

File: compiler/packages/eslint-plugin-react-compiler/src/rules/ReactCompilerRule.ts (hunk 2 above)
Comment (josephsavona) [reply compiler/packages/eslint-plugin-react-compiler/src/rules/ReactCompilerRule.ts:169]: Ohhh i've wanted that a few times

**Review by josephsavona — APPROVED** [review]

**Comment by poteto** [comment]
Sorry about not chiming in earlier, but I'm not sure I'm following that the fix is to truncate the error location to just the first line. Shouldn't we audit the locations we report instead and ensure that they're scoped to a more specific point in the source (not the component/hook itself)? 

In any case, this approach doesn't seem to play well with our internal eslint infra which expects that lint message locations fall within the bounds of the actual source code. I think it'd be better to revert this change and then find the errors that span the whole function and fix their reported locations

**Pull request merged**